Extract HackyAI logic into *Module traits - part 1

reaperrr opened the pull request:
This allows to disable or replace them at will.

This first PR turns the already existing harvester manager and support power manager into player traits, and directly extracts the capture logic and reaction-to-damage logic to traits as well.

The yaml changes were originally 100% update rule output, but I removed redundancies later (most AIs in each mod used the exact same SupportPower settings).

The only module I haven't really tested yet is the capture module, due to lack of testcases in our shipping mods.

Other than that, everything including the update rule has been tested and should work as intended.

reaperrr committed:
Extract AICaptureModule from HackyAI

Another step towards making the AI more modular and HackyAI.cs less monolithic.

reaperrr committed:
Change AIHarvesterManager into *Module trait

Makes it easier to replace, disable or override AI harvester management.

reaperrr committed:
Change AISupportPowerManager to *Module player trait

reaperrr committed:
Extract AIDamageReactionModule from HackyAI

reaperrr committed:
Update rule for first batch of AI modules

GraionDilach commented on OpenRA.Mods.Common/UpdateRules/UpdatePath.cs:
```diff
@@ -86,7 +86,7 @@ public class UpdatePath
 
 			new UpdatePath("playtest-20180729", new UpdateRule[]
 			{
-				// Bleed only changes here
```
Think we want to keep this comment.

reaperrr replied:
Oops, yes.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -270,14 +236,17 @@ public CPos GetRandomBaseCenter()
 
-		AIHarvesterManager harvManager;
-		AISupportPowerManager supportPowerManager;
+		AIHarvesterModule harvModule;
```
Extracting these to modules but then keeping the module types hardcoded hides a lot of the potential benefits here.  Would it be feasible to have a single generic `IAIModule` interface and then define a `Dictionary<string, string>` in the `HackyAIInfo` for condition: enable module name?  Or are we still relying too much on calling into the internal details of the classes?  If we are, then please lets use interfaces for each type.

reaperrr replied:
> Or are we still relying too much on calling into the internal details of the classes?

I'm afraid so, but I'll check if I can reduce that further.

reaperrr commented:
> Is it just me, or the SupportPower modules in TD/RA are defined wrong - you seem to aim for sharing them between the AIs but then each AI's using their own name to assign them.

Look at the respective "Remove..." commits, they fix that.
The first yaml commit is just the unmodified update rule output. It would be way too complicated to check for duplication during the update rule, so I just extract a support power module per AI.
The 2nd commit for each mod then removes the redundancy manually.

GraionDilach commented:
Ahyeah, I can see now - D2k is the bugged one.

Gerseras commented:
reaperr are you still working on this? If so are there ways to help you as i would like to have the ai modular so we can implement a own/custom basebuilder module for the kknd mods so the ai learns to handle the other way of constructing buildings

reaperrr commented:
> reaperr are you still working on this? If so are there ways to help you as i would like to have the ai modular so we can implement a own/custom basebuilder module for the kknd mods so the ai learns to handle the other way of constructing buildings

I've been working mostly on other things in the meantime, but I think I can whip this into good-enough shape relatively quickly.

reaperrr commented:
Rebased and updated.

@Gerseras @IceReaper Feel free to help reviewing/testing now :)

Locally I have the BaseManagerModule ready as well (except for the update rule and yaml changes, but I guess this probably won't get merged immediately anyway).

Gerseras commented:
@reaperrr We will take a look at it next week and try to use a custom HarvesterModule for our Tanker logic
if that works we would love to have the basebuilder module as that could become quite intresting

pchote commented:
We have been a bit inconsistent with "AI" vs "Bot" before now - we have tended to use "Bot" in the code, and "AI" for user facing things, except for `HackyAI`.  IMO it would be better to call these `IBotModule`s as this is a bit more specific than "AI", which could also refer to unit-level AI on player-owned units.

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -162,8 +162,8 @@ public class BuildingCategories
-		[Desc("Avoid enemy actors nearby when searching for a new resource patch. Should be somewhere near the max weapon range.")]
-		public readonly WDist HarvesterEnemyAvoidanceRadius = WDist.FromCells(8);
+		[Desc("Which harvester handling module to use. Leave empty to disable.")]
+		public readonly string HarvesterModule = "default-harvester-module";
```
Having specifically behaving modules referred to via strings like this sets off warning bells in my head.  What if someone references a support power module type here, or something different?

If we want to refer to specific types of module here, then we should come up with a way to annotate them - perhaps using a new attribute with a tag type.

A much better fix would be to not have specific `*Modules` at all, and instead a generic `Modules` list, with `HackyAI` itself not caring about what each module is/does.

Phrohdoh commented on OpenRA.Mods.Common/AI/AIDamageReactionModule.cs:
```diff
@@ -0,0 +1,121 @@
+	public class AIDamageReactionModuleInfo : IAIModuleInfo
+	{
+		[Desc("Name for identification purposes.")]
+		public readonly string Name = "default-damagereaction-module";
```
Why is `default` in this name?

Phrohdoh replied:
If we're going to have modules added by default shouldn't the harvester one also be a default?

reaperrr replied:
That was a relic from earlier code revisions.

Phrohdoh commented on OpenRA.Mods.Common/AI/AIHarvesterModule.cs:
```diff
@@ -20,54 +20,52 @@
 	{
-		readonly HackyAI ai;
+		[Desc("Name for identification purposes.")]
+		public readonly string Name = "default-harvester-module";
```
IIRC the other PR doesn't have `default-` in the harvester module's name, which do we want?

reaperrr replied:
The plumbing has out-evolved this Name property in the split off PRs

Phrohdoh commented on mods/d2k/rules/ai.yaml:
```diff
@@ -332,6 +250,9 @@ Player:
+	AIHarvesterModule:
+	AIDamageReactionModule:
+	AISupportPowerModule:
 		SupportPowerDecisions:
```
Nit: Could we rename this property? No need, IMO, to have `SupportPower` duplicated now that this lives under a support-power-specific trait.

reaperrr replied:
Will do that in #15786.

reaperrr commented:
This is both too monolithic and too stale by now, I'll split off the modules in separate PRs. Closing.
